Add broker_channel_error_retry option

nkns165 opened the pull request:
## Description

<!-- Please describe your pull request.

NOTE: All patches should be made against master, not a maintenance branch like
3.1, 2.5, etc.  That is unless the bug is already fixed in master, but not in
that version series.

If it fixes a bug or resolves a feature request,
be sure to link to that issue via (Fixes #4412) for example.
-->
Resolves #7946

As the error from Redis failover was `ResponseError` which is categorized in `channel_errors` (not `connection_errors`), connection retry didn't happen.
I have added a new option `broker_channel_error_retry` which also retries for `channel_errors`. Its default value is `False` so that existing behavior is not changed.

Sorry that the actual fix is a bit different from what I described in the issue.

auvipy reviewed (changes requested):
it seems all integration tests are failing, can you please investigate? also we expect new features to be atleast unit tested

nkns165 commented:
@auvipy Thank you for checking, I'd like to try adding unit tests.

But I have a question. I've tried to build a Docker container according to [this document](https://docs.celeryq.dev/en/master/contributing.html#developing-and-testing-with-docker), but even on the master branch it fails during pip install.

<details>
<summary> error log </summary>

```
#0 31.05 ERROR: Cannot install billiard 4.0.2 (from git+https://github.com/celery/billiard.git), kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git) and pytest-celery because these package versions have conflicting dependencies.
#0 31.05 
#0 31.05 The conflict is caused by:
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.7 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.6 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.4 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.3 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.2 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.1 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.2.0 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.1.2 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.1.1 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.1.0 depends on billiard<4.0 and >=3.6.4.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.5 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.4 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.3 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.2 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.1 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 5.0.0 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 4.4.7 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     celery 4.4.6 depends on kombu<4.7 and >=4.6.10
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     celery 4.4.5 depends on kombu<4.7 and >=4.6.10
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 4.4.4 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 4.4.3 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 4.4.2 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested billiard 4.0.2 (from git+https://github.com/celery/billiard.git)
#0 31.05     celery 4.4.1 depends on billiard<4.0 and >=3.6.3.0
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     The user requested kombu 5.3.0b2 (from git+https://github.com/celery/kombu.git)
#0 31.05     celery 4.4.0 depends on kombu<4.7 and >=4.6.7
#0 31.05 
#0 31.05 To fix this you could try to:
#0 31.05 1. loosen the range of package versions you've specified
#0 31.05 2. remove package versions to allow pip attempt to solve the dependency conflict
#0 31.05 
#0 31.05 ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

</details>

Also, other people's pull requests also fail in integration tests.
- https://github.com/celery/celery/pull/7952
- https://github.com/celery/celery/pull/7945

Failing situation seems the same (billiard/kombu is related)
- This pull request's Celery / Integration (3.7, redis):  https://github.com/celery/celery/actions/runs/3658861781/jobs/6184638126
- Same tests for https://github.com/celery/celery/pull/7952: https://github.com/celery/celery/actions/runs/3664845135/jobs/6195639079

Are these tests working on the master branch?

auvipy commented:
yes, we need to figure out and fix them. also the docker is outdated already I think

auvipy commented:
the failures are mostly network and CI related issues not actual broken tests as it seems to me from the logs

nkns165 commented:
Thanks, I will focus on adding a unit test. Sorry to be late, I still have some trouble in executing unit tests on local.

auvipy commented on docs/userguide/configuration.rst:
```diff
@@ -2806,6 +2806,17 @@ to the AMQP broker.
 If this is set to :const:`0` or :const:`None`, we'll retry forever.
 
+``broker_channel_error_retry``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
we should also add version added 5.3 annotation here

nkns165 replied:
thanks, added in [0b985dc](https://github.com/celery/celery/pull/7951/commits/0b985dce930ad20abdd3ccb9ee806764515d7f26)

auvipy reviewed (changes requested):
please pull from master once again.

Nusnus commented:
> Thanks, I will focus on adding a unit test. Sorry to be late, I still have some trouble in executing unit tests on local.

What troubles do you have with local testing? Maybe I can assist :)
Also does running the tests with `tox` in the terminal works? Before setting up testing inside an IDE.
e.g. Running `tox -e 3.9-integration` from the CLI

nkns165 commented:
> > Thanks, I will focus on adding a unit test. Sorry to be late, I still have some trouble in executing unit tests on local.
> 
> What troubles do you have with local testing? Maybe I can assist :) 

@Nusnus 
Thank you, I have resolved the first problem I was facing.

Previously I had dependency conflict around billiard, but it was because my virtualenv contained celery 5.2.7 by mistake. After I install the latest code locally by `pip install`, it was resolved and I was able to run unit tests.
Now,
- I can make unit tests pass at least for files I modified by `pytest t/unit/worker/test_consumer.py`.
- But `pytest t/unit` causes some failures/skips. It's same on master branch, so if no problem I'd like to leave it.
```
======================= 2557 passed, 48 skipped, 3 xfailed, 45 warnings, 127 subtests passed in 67.84s (0:01:07) =======================
```

<details>
<summary>Failed tests</summary>

```
=========================================================== warnings summary ===========================================================
t/unit/app/test_backends.py: 1 warning
t/unit/contrib/test_worker.py: 2 warnings
t/unit/worker/test_consumer.py: 2 warnings
t/unit/worker/test_worker.py: 17 warnings
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/worker/consumer/consumer.py:498: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
  whether broker connection retries are made during startup in Celery 6.0 and above.
  If you wish to retain the existing behavior for retrying connections on startup,
  you should set broker_connection_retry_on_startup to True.
    warnings.warn(

t/unit/fixups/test_django.py::test_DjangoWorkerFixup::test_validate_models
  /Users/hiroko.tamagawa/workspace/celery/t/unit/fixups/test_django.py:275: PytestWarning: Value of environment variable CELERY_SKIP_CHECKS type should be str, but got True (type: bool); converted to str implicitly
    patching.setenv('CELERY_SKIP_CHECKS', True)

t/unit/tasks/test_canvas.py::test_Signature::test_callback_stamping
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/canvas.py:422: AlwaysEagerIgnored: task_always_eager has no effect on send_task
    return _apply(args, kwargs, **options)

t/unit/tasks/test_canvas.py::test_chord::test_links_to_body
t/unit/tasks/test_canvas.py::test_chord::test_flag_allow_error_cb_on_chord_header
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/canvas.py:2258: CPendingDeprecationWarning: task_allow_error_cb_on_chord_header=False is pending deprecation in a future release of Celery.
  Please test the new behavior by setting task_allow_error_cb_on_chord_header to True and report any concerns you might have in our issue tracker before we make a final decision regarding how errbacks should behave when used with chords.
    warnings.warn(

t/unit/tasks/test_trace.py::test_trace::test_backend_error_should_report_failure
t/unit/worker/test_request.py::test_Request::test_execute_backend_error_acks_late
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py:659: RuntimeWarning: Exception raised outside body: Exception():
  Traceback (most recent call last):
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 517, in trace_task
      task.backend.mark_as_done(
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1093, in __call__
      return self._mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1097, in _mock_call
      return self._execute_mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1152, in _execute_mock_call
      raise effect
  Exception
  
    warn(RuntimeWarning(

t/unit/tasks/test_trace.py::test_trace::test_backend_error_should_report_failure
t/unit/worker/test_request.py::test_Request::test_execute_backend_error_acks_late
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py:659: RuntimeWarning: Exception raised outside body: Exception():
  Traceback (most recent call last):
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 517, in trace_task
      task.backend.mark_as_done(
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1093, in __call__
      return self._mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1097, in _mock_call
      return self._execute_mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1152, in _execute_mock_call
      raise effect
  Exception
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 583, in trace_task
      return task.__trace__(uuid, args, kwargs, request)
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 571, in trace_task
      I, _, _, _ = on_error(task_request, exc)
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 378, in on_error
      R = I.handle_error_state(
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 175, in handle_error_state
      return {
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 222, in handle_failure
      task.backend.mark_as_failure(
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1093, in __call__
      return self._mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1097, in _mock_call
      return self._execute_mock_call(*args, **kwargs)
    File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/unittest/mock.py", line 1152, in _execute_mock_call
      raise effect
  Exception
  
    warn(RuntimeWarning(

t/unit/worker/test_consumer.py: 10 warnings
t/unit/worker/test_worker.py: 1 warning
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/worker/consumer/consumer.py:386: CPendingDeprecationWarning: 
  In Celery 5.1 we introduced an optional breaking change which
  on connection loss cancels all currently executed tasks with late acknowledgement enabled.
  These tasks cannot be acknowledged as the connection is gone, and the tasks are automatically redelivered back to the queue.
  You can enable this behavior using the worker_cancel_long_running_tasks_on_connection_loss setting.
  In Celery 5.1 it is set to False by default. The setting will be set to True by default in Celery 6.0.
  
    warnings.warn(CANCEL_TASKS_BY_DEFAULT, CPendingDeprecationWarning)

t/unit/worker/test_consumer.py::test_Consumer::test_ensure_connected[None-False]
t/unit/worker/test_worker.py::test_Consumer::test_receive_message_eta
t/unit/worker/test_worker.py::test_Consumer::test_receive_message_eta
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/worker/consumer/consumer.py:498: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
  whether broker connection retries are made during startup in Celery 6.0 and above.
  If you wish to retain the existing behavior for retrying connections on startup,
  you should set broker_connection_retry_on_startup to False.
    warnings.warn(

t/unit/worker/test_request.py::test_Request::test_from_message_invalid_kwargs
  /Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py:659: RuntimeWarning: Exception raised outside body: InvalidTaskError('Task keyword arguments is not a mapping'):
  Traceback (most recent call last):
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 400, in trace_task
      kwargs.items
  AttributeError: 'str' object has no attribute 'items'
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/hiroko.tamagawa/.virtualenvs/celery/lib/python3.9/site-packages/celery/app/trace.py", line 402, in trace_task
      raise InvalidTaskError(
  celery.exceptions.InvalidTaskError: Task keyword arguments is not a mapping
  
    warn(RuntimeWarning(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

</details>

> Also does running the tests with `tox` in the terminal works? Before setting up testing inside an IDE. e.g. Running `tox -e 3.9-integration` from the CLI

Sorry I don't know much about `tox`, is this set up correct?

```
pip install tox
pip install tox-gh-actions
```

After that I was able to run some tests by `tox -e 3.9-integration`, but the result was same for master and this feature branch.

<details>
<summary>Log</summary>

```
/Users/hiroko.tamagawa/workspace/celery/.tox/3.9-integration/lib/python3.9/site-packages/setuptools/config/setupcfg.py:508: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
  warnings.warn(msg, warning_class)
3.9-integration develop-inst-noop: /Users/hiroko.tamagawa/workspace/celery
3.9-integration installed: alabaster==0.7.12,amqp==5.1.1,async-timeout==4.0.2,attrs==22.1.0,azure-core==1.26.1,azure-storage-blob==12.14.1,Babel==2.11.0,billiard==4.1.0,boto3==1.26.30,botocore==1.29.30,bump2version==1.0.1,bumpversion==0.6.0,cachetools==5.2.0,cassandra-driver==3.25.0,-e git+ssh://git@github.com/celery/celery.git@dd811b37717635b5f7151a7adf9f5bf12e1bc0c6#egg=celery,certifi==2022.12.7,cffi==1.15.1,cfgv==3.3.1,chardet==5.1.0,charset-normalizer==2.1.1,click==8.1.3,click-didyoumean==0.3.0,click-plugins==1.1.1,click-repl==0.2.0,codecov==2.1.12,colorama==0.4.6,couchbase==4.1.1,coverage==6.5.0,cryptography==38.0.4,DateTime==4.7,distlib==0.3.6,dnspython==2.2.1,docutils==0.19,elasticsearch==7.17.8,ephem==4.1.3,eventlet==0.33.2,exceptiongroup==1.0.4,filelock==3.8.2,flake8==6.0.0,flake8-docstrings==1.6.0,flakeplus==1.1.0,future==0.18.2,geomet==0.2.1.post1,gevent==22.10.2,greenlet==2.0.1,identify==2.5.9,idna==3.4,imagesize==1.4.1,importlib-metadata==5.1.0,iniconfig==1.1.1,isodate==0.6.1,Jinja2==3.1.2,jmespath==1.0.1,kombu @ git+https://github.com/celery/kombu.git@9ee6fdea7d4712a0e0584e9410a5954e42a8d1b0,livereload==2.6.3,MarkupSafe==2.1.1,mccabe==0.7.0,mock==4.0.3,moto==4.0.11,msgpack==1.0.4,msrest==0.7.1,mypy==0.991,mypy-extensions==0.4.3,nodeenv==1.7.0,oauthlib==3.2.2,packaging==22.0,platformdirs==2.6.0,pluggy==1.0.0,pre-commit==2.20.0,prompt-toolkit==3.0.36,pyArango==2.0.1,pycodestyle==2.10.0,pycouchdb==1.14.2,pycparser==2.21,pycurl==7.43.0.5,pydocstyle==6.1.1,pydocumentdb==2.3.5,pyflakes==3.0.1,Pygments==2.13.0,pylibmc==1.6.3,pymongo==4.3.3,pyproject_api==1.2.1,pytest==7.2.0,pytest-celery==0.0.0,pytest-click==1.1.0,pytest-cov==4.0.0,pytest-github-actions-annotate-failures==0.1.7,pytest-order==1.0.1,pytest-rerunfailures==10.3,pytest-subtests==0.9.0,pytest-timeout==2.1.0,python-consul2==0.1.5,python-dateutil==2.8.2,python-memcached==1.59,pytz==2022.6,PyYAML==6.0,redis==4.4.0,requests==2.28.1,requests-oauthlib==1.3.1,responses==0.22.0,s3transfer==0.6.0,six==1.16.0,snowballstemmer==2.2.0,softlayer-messaging==1.0.3,Sphinx==5.3.0,sphinx-autobuild==2021.3.14,sphinx-celery==2.0.0,sphinx-click==4.4.0,sphinx-testing==1.0.1,sphinx2rst==1.1.0,sphinxcontrib-applehelp==1.0.2,sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==2.0.0,sphinxcontrib-jsmath==1.0.1,sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.5,SQLAlchemy==1.4.45,tblib==1.7.0,toml==0.10.2,tomli==2.0.1,tornado==6.2,tox==4.0.10,types-toml==0.10.8.1,typing_extensions==4.4.0,Unipath==1.1,urllib3==1.26.13,vine==5.0.0,virtualenv==20.17.1,wcwidth==0.2.5,Werkzeug==2.2.2,xmltodict==0.13.0,zipp==3.11.0,zope.event==4.5.0,zope.interface==5.5.2
3.9-integration run-test-pre: PYTHONHASHSEED='3632015361'
3.9-integration run-test: commands[0] | pytest -xsv t/integration
========================================================= test session starts ==========================================================
platform darwin -- Python 3.9.6, pytest-7.2.0, pluggy-1.0.0 -- /Users/hiroko.tamagawa/workspace/celery/.tox/3.9-integration/bin/python
cachedir: .tox/3.9-integration/.pytest_cache
rootdir: /Users/hiroko.tamagawa/workspace/celery, configfile: pyproject.toml
plugins: subtests-0.9.0, github-actions-annotate-failures-0.1.7, rerunfailures-10.3, click-1.1.0, timeout-2.1.0, order-1.0.1, cov-4.0.0, celery-0.0.0
collected 206 items                                                                                                                    

t/integration/test_backend.py::test_AzureBlockBlobBackend::test_crud SKIPPED (Environment variable AZUREBLOCKBLOB_URL required)
t/integration/test_backend.py::test_AzureBlockBlobBackend::test_get_missing SKIPPED (Environment variable AZUREBLOCKBLOB_URL...)
t/integration/test_canvas.py::test_link_error::test_link_error_eager PASSED
t/integration/test_canvas.py::test_link_error::test_link_error PASSED
t/integration/test_canvas.py::test_link_error::test_link_error_callback_error_callback_retries_eager PASSED
t/integration/test_canvas.py::test_link_error::test_link_error_callback_retries FAILED

=============================================================== FAILURES ===============================================================
___________________________________________ test_link_error.test_link_error_callback_retries ___________________________________________

self = <promise: 0x10cd21940>

    def __call__(self):
        try:
>           return self.__value__
E           AttributeError: 'ChannelPromise' object has no attribute '__value__'

.tox/3.9-integration/lib/python3.9/site-packages/kombu/utils/functional.py:32: AttributeError

During handling of the above exception, another exception occurred:

self = <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>, ConnectionError = <class 'kombu.exceptions.OperationalError'>
ChannelError = <class 'kombu.exceptions.OperationalError'>

    @contextmanager
    def _reraise_as_library_errors(
            self,
            ConnectionError=exceptions.OperationalError,
            ChannelError=exceptions.OperationalError):
        try:
>           yield

.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:461: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>, errback = None, max_retries = None, interval_start = 2
interval_step = 2, interval_max = 30, callback = None, reraise_as_library_errors = True, timeout = 4

    def _ensure_connection(
        self, errback=None, max_retries=None,
        interval_start=2, interval_step=2, interval_max=30,
        callback=None, reraise_as_library_errors=True,
        timeout=None
    ):
        """Ensure we have a connection to the server.
    
        If not retry establishing the connection with the settings
        specified.
    
        Arguments:
            errback (Callable): Optional callback called each time the
                connection can't be established.  Arguments provided are
                the exception raised and the interval that will be
                slept ``(exc, interval)``.
    
            max_retries (int): Maximum number of times to retry.
                If this limit is exceeded the connection error
                will be re-raised.
    
            interval_start (float): The number of seconds we start
                sleeping for.
            interval_step (float): How many seconds added to the interval
                for each retry.
            interval_max (float): Maximum number of seconds to sleep between
                each retry.
            callback (Callable): Optional callback that is called for every
                internal iteration (1 s).
            timeout (int): Maximum amount of time in seconds to spend
                attempting to connect, total over all retries.
        """
        if self.connected:
            return self._connection
    
        def on_error(exc, intervals, retries, interval=0):
            round = self.completes_cycle(retries)
            if round:
                interval = next(intervals)
            if errback:
                errback(exc, interval)
            self.maybe_switch_next()  # select next host
    
            return interval if round else 0
    
        ctx = self._reraise_as_library_errors
        if not reraise_as_library_errors:
            ctx = self._dummy_context
        with ctx():
>           return retry_over_time(
                self._connection_factory, self.recoverable_connection_errors,
                (), {}, on_error, max_retries,
                interval_start, interval_step, interval_max,
                callback, timeout=timeout
            )

.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:448: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fun = <bound method Connection._connection_factory of <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>>
catch = (<class 'amqp.exceptions.RecoverableConnectionError'>, <class 'OSError'>, <class 'OSError'>, <class 'OSError'>), args = []
kwargs = {}, errback = <function Connection._ensure_connection.<locals>.on_error at 0x10dc0c940>, max_retries = None, interval_start = 2
interval_step = 2, interval_max = 30, callback = None, timeout = 4

    def retry_over_time(fun, catch, args=None, kwargs=None, errback=None,
                        max_retries=None, interval_start=2, interval_step=2,
                        interval_max=30, callback=None, timeout=None):
        """Retry the function over and over until max retries is exceeded.
    
        For each retry we sleep a for a while before we try again, this interval
        is increased for every retry until the max seconds is reached.
    
        Arguments:
            fun (Callable): The function to try
            catch (Tuple[BaseException]): Exceptions to catch, can be either
                tuple or a single exception class.
    
        Keyword Arguments:
            args (Tuple): Positional arguments passed on to the function.
            kwargs (Dict): Keyword arguments passed on to the function.
            errback (Callable): Callback for when an exception in ``catch``
                is raised.  The callback must take three arguments:
                ``exc``, ``interval_range`` and ``retries``, where ``exc``
                is the exception instance, ``interval_range`` is an iterator
                which return the time in seconds to sleep next, and ``retries``
                is the number of previous retries.
            max_retries (int): Maximum number of retries before we give up.
                If neither of this and timeout is set, we will retry forever.
                If one of this and timeout is reached, stop.
            interval_start (float): How long (in seconds) we start sleeping
                between retries.
            interval_step (float): By how much the interval is increased for
                each retry.
            interval_max (float): Maximum number of seconds to sleep
                between retries.
            timeout (int): Maximum seconds waiting before we give up.
        """
        kwargs = {} if not kwargs else kwargs
        args = [] if not args else args
        interval_range = fxrange(interval_start,
                                 interval_max + interval_start,
                                 interval_step, repeatlast=True)
        end = time() + timeout if timeout else None
        for retries in count():
            try:
>               return fun(*args, **kwargs)

.tox/3.9-integration/lib/python3.9/site-packages/kombu/utils/functional.py:314: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>

    def _connection_factory(self):
        self.declared_entities.clear()
        self._default_channel = None
>       self._connection = self._establish_connection()

.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:900: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>

    def _establish_connection(self):
        self._debug('establishing connection...')
>       conn = self.transport.establish_connection()

.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:827: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <kombu.transport.pyamqp.Transport object at 0x10c85acd0>

    def establish_connection(self):
        """Establish connection to the AMQP broker."""
        conninfo = self.client
        for name, default_value in self.default_connection_params.items():
            if not getattr(conninfo, name, None):
                setattr(conninfo, name, default_value)
        if conninfo.hostname == 'localhost':
            conninfo.hostname = '127.0.0.1'
        # when server_hostname is None, use hostname from URI.
        if isinstance(conninfo.ssl, dict) and \
                'server_hostname' in conninfo.ssl and \
                conninfo.ssl['server_hostname'] is None:
            conninfo.ssl['server_hostname'] = conninfo.hostname
        opts = dict({
            'host': conninfo.host,
            'userid': conninfo.userid,
            'password': conninfo.password,
            'login_method': conninfo.login_method,
            'virtual_host': conninfo.virtual_host,
            'insist': conninfo.insist,
            'ssl': conninfo.ssl,
            'connect_timeout': conninfo.connect_timeout,
            'heartbeat': conninfo.heartbeat,
        }, **conninfo.transport_options or {})
        conn = self.Connection(**opts)
        conn.client = self.client
>       conn.connect()

.tox/3.9-integration/lib/python3.9/site-packages/kombu/transport/pyamqp.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <AMQP Connection: 127.0.0.1:5672// (disconnected) at 0x10db61c10>, callback = None

    def connect(self, callback=None):
        # Let the transport.py module setup the actual
        # socket connection to the broker.
        #
        if self.connected:
            return callback() if callback else None
        try:
            self.transport = self.Transport(
                self.host, self.connect_timeout, self.ssl,
                self.read_timeout, self.write_timeout,
                socket_settings=self.socket_settings,
            )
>           self.transport.connect()

.tox/3.9-integration/lib/python3.9/site-packages/amqp/connection.py:323: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <TCPTransport: (disconnected) at 0x10cbc4cc0>

    def connect(self):
        try:
            # are we already connected?
            if self.connected:
                return
>           self._connect(self.host, self.port, self.connect_timeout)

.tox/3.9-integration/lib/python3.9/site-packages/amqp/transport.py:129: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <TCPTransport: (disconnected) at 0x10cbc4cc0>, host = '127.0.0.1', port = 5672, timeout = 4

    def _connect(self, host, port, timeout):
        entries = socket.getaddrinfo(
            host, port, socket.AF_UNSPEC, socket.SOCK_STREAM, SOL_TCP,
        )
        for i, res in enumerate(entries):
            af, socktype, proto, canonname, sa = res
            try:
                self.sock = socket.socket(af, socktype, proto)
                try:
                    set_cloexec(self.sock, True)
                except NotImplementedError:
                    pass
                self.sock.settimeout(timeout)
>               self.sock.connect(sa)
E               ConnectionRefusedError: [Errno 61] Connection refused

.tox/3.9-integration/lib/python3.9/site-packages/amqp/transport.py:184: ConnectionRefusedError

The above exception was the direct cause of the following exception:

self = <t.integration.test_canvas.test_link_error object at 0x10bf3fdc0>

    @pytest.mark.xfail(raises=TimeoutError, reason="Task is timeout instead of returning exception")
    def test_link_error_callback_retries(self):
        exception = ExpectedException("Task expected to fail", "test")
>       result = fail.apply_async(
            args=("test",),
            link_error=retry_once.s(countdown=None)
        )

t/integration/test_canvas.py:159: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
celery/app/task.py:588: in apply_async
    return app.send_task(
celery/app/base.py:794: in send_task
    amqp.send_task_message(P, name, message, **options)
celery/app/amqp.py:517: in send_task_message
    ret = producer.publish(
.tox/3.9-integration/lib/python3.9/site-packages/kombu/messaging.py:183: in publish
    return _publish(
.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:538: in _ensured
    return fun(*args, **kwargs)
.tox/3.9-integration/lib/python3.9/site-packages/kombu/messaging.py:192: in _publish
    channel = self.channel
.tox/3.9-integration/lib/python3.9/site-packages/kombu/messaging.py:215: in _get_channel
    channel = self._channel = channel()
.tox/3.9-integration/lib/python3.9/site-packages/kombu/utils/functional.py:34: in __call__
    value = self.__value__ = self.__contract__()
.tox/3.9-integration/lib/python3.9/site-packages/kombu/messaging.py:231: in <lambda>
    channel = ChannelPromise(lambda: connection.default_channel)
.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:918: in default_channel
    self._ensure_connection(**conn_opts)
.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:448: in _ensure_connection
    return retry_over_time(
/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/contextlib.py:135: in __exit__
    self.gen.throw(type, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Connection: amqp://guest:**@127.0.0.1:5672// at 0x10ccc6340>, ConnectionError = <class 'kombu.exceptions.OperationalError'>
ChannelError = <class 'kombu.exceptions.OperationalError'>

    @contextmanager
    def _reraise_as_library_errors(
            self,
            ConnectionError=exceptions.OperationalError,
            ChannelError=exceptions.OperationalError):
        try:
            yield
        except (ConnectionError, ChannelError):
            raise
        except self.recoverable_connection_errors as exc:
>           raise ConnectionError(str(exc)) from exc
E           kombu.exceptions.OperationalError: [Errno 61] Connection refused

.tox/3.9-integration/lib/python3.9/site-packages/kombu/connection.py:465: OperationalError
---------------------------------------------------------- Captured log call -----------------------------------------------------------
WARNING  kombu.connection:connection.py:644 No hostname was supplied. Reverting to default 'localhost'
======================================================= short test summary info ========================================================
FAILED t/integration/test_canvas.py::test_link_error::test_link_error_callback_retries - kombu.exceptions.OperationalError: [Errno 61] Connection refused
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================ 1 failed, 3 passed, 2 skipped in 8.38s ================================================
ERROR: InvocationError for command /Users/hiroko.tamagawa/workspace/celery/.tox/3.9-integration/bin/pytest -xsv t/integration (exited with code 1)
_______________________________________________________________ summary ________________________________________________________________
ERROR:   3.9-integration: commands failed
```

</details>

Is there anything else I should do before requesting the review again?

Actuallyl I don't know how to fix the error by `pre-commit.ci`...

auvipy commented:
you can skip the warnings those are not related to this PR. only need to fix the Integrations test failure if any

nkns165 commented:
@auvipy Thanks! integration tests looks fine, so re-requested review.
